### PR TITLE
fix: Fixed SAR model for training and inference in PyTorch

### PR DESCRIPTION
**Pull request by fg-mindee**
Following up on #802, this PR introduces the following changes:
- fixes a major typo in the loss computation (the sequence masking was inverted)
- cleaned encoder and decoder for a much more understandable version

I tried training the model with the synthetic training set, and a real-world validation set and managed to reach almost 100% of exact match after a few epochs.

Any feedback is welcome!

### CODE REVIEW

**Comment by felixdittrich92** [comment]
@fg-mindee 
Offtopic:
I have faiced the same in my tests (~99% exact) but after loading this trained model in the whole OCR pipeline there was only wrong results 😅 Do you tested this also ? 😄  (train size was in my test cases 1M / val ~ 200k)

**Review by charlesmindee — APPROVED** [review]
Thanks!

**Comment by fg-mindee** [comment]
> @fg-mindee Offtopic: I have faiced the same in my tests (~99% exact) but after loading this trained model in the whole OCR pipeline there was only wrong results sweat_smile Do you tested this also ? smile (train size was in my test cases 1M / val ~ 200k)

I had the same issue until I changed the transformation pipeline during training. On my end, I used a real-world validation dataset, and it worked! If it doesn't on the full OCR pipeline, the problem comes from something else then :)
